collect_app/src/test/java/org/odk/collect/android/external/ExternalDataReaderTest.java

```diff
@@ -0,0 +1,225 @@
+                COLUMN_DATASET_FILENAME,
+                CustomSQLiteQueryBuilder.quoteStringLiteral(SIMPLE_SEARCH_EXTERNAL_CSV_FILENAME));
+        Cursor cursor = db.query(testMetadataTable, columnNames, selectCriteria, null, null, null, null);
+        cursor.moveToFirst();
+        long fileTimestamp = cursor.getLong(0);
+
+        assertThat(fileTimestamp, is(csvFile.lastModified()));
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfDatabaseFileIsDeleted() {
+        // Create the DB file with an initial import
+        ExternalDataReader externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+
+        dbFile.delete();
+
+        // Reimport
+        externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfMetadataTableIsMissing() {
+        // Create the DB file with an initial import
+        ExternalDataReader externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+
+        // Remove the metadata table (mimicking prior versions without the metadata table)
+        SQLiteDatabase db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        SQLiteUtils.dropTable(db, EXTERNAL_METADATA_TABLE_NAME);
+        db.close();
+
+        // Reimport
+        externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+        db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        assertThat("metadata table should be recreated", SQLiteUtils.doesTableExist(db, EXTERNAL_METADATA_TABLE_NAME));
+        db.close();
+    }
+
+    @Test
+    public void doImport_reimportsCsvIfFileIsUpdated() throws IOException, InterruptedException {
+        // Create the DB file with an initial import
+        ExternalDataReader externalDataReader = new ExternalDataReaderImpl(null);
+        externalDataReader.doImport(formDefToCsvMedia);
+        assertThat(dbFile.exists(), is(true));
+
+        SQLiteDatabase db = SQLiteDatabase.openDatabase(dbFile.getAbsolutePath(), null, SQLiteDatabase.OPEN_READWRITE);
+        assertThat(db.rawQuery(SELECT_ALL_DATA_QUERY, null).getCount(), is(3));
+
+        long originalTimestamp = csvFile.lastModified();
+        long metadataLastModified = ExternalSQLiteOpenHelper.getLastImportTimestamp(db, EXTERNAL_METADATA_TABLE_NAME, csvFile);
+        assertThat(metadataLastModified, is(originalTimestamp));
+
+        Thread.sleep(1000); // this test is so fast that we need to guarantee the modified time changes
```

Comment:
I think there is a way to fake this with fake file layer. Can't remember the name of the framework though. We can stick with this right now as I'm pretty sure it's still deterministic.